### PR TITLE
Add `parseFlakeRef` and `flakeRefToString` builtins

### PR DESCRIPTION
**Pull request by aakropotkin**
# Motivation

Over the last year or so I've run into several use cases where I need to parse and/or serialize URLs for use by `builtins.fetchTree` or `builtins.getFlake`, largely in order to produce _lockfile-like_ files for lang2nix frameworks or tools which use `nix` internally to drive builds.

I've gone through the painstaking process of emulating `nix::FlakeRef::fromAttrs` and `nix::parseFlakeRef` several times with mixed success; but these are difficult to create and even harder to maintain if I hope to stay aligned with changes to the real parser/serializer.

# Context

I understand why adding new `builtins` isn't something we want to do flagrantly. I'm recommending this addition simply because I keep encountering use cases where I need to parse/serialize these URIs in `nix` expressions, and I want a reliable solution.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libexpr/flake/flake.cc [hunk 1]
```diff
@@ -793,6 +793,104 @@ static RegisterPrimOp r2({
...
+                              std::string(attr.value->str()));
+                break;
+            case nThunk:
+            case nFloat:
+            case nPath:
+            case nNull:
+            case nAttrs:
+            case nList:
+            case nFunction:
+            case nExternal:
```
Comment (edolstra) [inline src/libexpr/flake/flake.cc:863]:
```suggestion
```

Seems unnecessary given the default case.

File: src/libexpr/flake/flake.cc [hunk 2]
```diff
@@ -793,6 +793,104 @@ static RegisterPrimOp r2({
...
+            default:
+                state.error(
```
Comment (edolstra) [inline src/libexpr/flake/flake.cc:865]: This error should be impossible right? So probably this should be an assertion failure.

File: src/libexpr/flake/flake.cc (hunk 1 above)
Comment (aakropotkin) [reply src/libexpr/flake/flake.cc:863]: I agree but GCC threw an error without them. Possibly an optional flag floating around the Makefile?

Comment (edolstra) [reply src/libexpr/flake/flake.cc:863]: I think you can add
```
#pragma GCC diagnostic ignored "-Wswitch-enum"
```
to the file. Or replace the switch with a series of if/then/elses.

Comment (aakropotkin) [reply src/libexpr/flake/flake.cc:863]: Will do

**Comment by aakropotkin** [comment]
@edolstra bump ❤️

File: src/libexpr/flake/flake.cc [hunk 3]
```diff
@@ -793,6 +793,99 @@ static RegisterPrimOp r2({
...
+        } else if (t == nBool) {
+            attrs.emplace(state.symbols[attr.name],
+                          Explicit<bool> { attr.value->boolean });
+        } else if (t == nString) {
+            attrs.emplace(state.symbols[attr.name],
+                          std::string(attr.value->str()));
+        } else {
+            state.error(
+                "flake-ref attr sets may only contain integers, booleans, "
+                "and strings, but attribute %s is %s",
```
Comment (edolstra) [inline src/libexpr/flake/flake.cc:863]:
```suggestion
                "flake reference attribute sets may only contain integers, Booleans, "
                "and strings, but attribute '%s' is %s",
```

File: src/libexpr/flake/flake.cc [hunk 4]
```diff
@@ -793,6 +793,99 @@ static RegisterPrimOp r2({
...
+    }
+    auto flakeRef = FlakeRef::fromAttrs(attrs);
+    v.mkString(flakeRef.to_string());
+}
+
+static RegisterPrimOp r4({
+    .name =  "__flakeRefToString",
+    .args = {"attrs"},
+    .doc = R"(
+      Stringize an exploded flake reference, and return its url form.
```
Comment (edolstra) [inline src/libexpr/flake/flake.cc:876]:
```suggestion
      Convert a flake reference from attribute set format to URL format.
```

File: src/libexpr/flake/flake.cc [hunk 5]
```diff
@@ -793,6 +793,99 @@ static RegisterPrimOp r2({
...
+        Value & vv = binds.alloc(s);
+        if (std::holds_alternative<std::string>(value)) {
+            vv.mkString(std::get<std::string>(value));
+        }
+        else if (std::holds_alternative<uint64_t>(value)) {
+            vv.mkInt(std::get<uint64_t>(value));
+        }
+        else { /* bool */
+            vv.mkBool(std::get<Explicit<bool>>(value).t);
+        }
```
Comment (Ericson2314) [inline src/libexpr/flake/flake.cc:817]: Please use `std::visit(overloaded { ... }, value)` so we get completeness checking.

File: src/libexpr/flake/flake.cc [hunk 6]
```diff
@@ -793,6 +793,99 @@ static RegisterPrimOp r2({
...
+    Value * * args,
+    Value & v)
+{
+    std::string flakeRefS(state.forceStringNoCtx(*args[0], pos,
+        "while evaluating the argument passed to builtins.parseFlakeRef"));
+    auto attrs = parseFlakeRef(flakeRefS, {}, true).toAttrs();
+    auto binds = state.buildBindings(attrs.size());
+    for ( const auto & [key, value] : attrs ) {
+        Symbol  s  = state.symbols.create(key);
+        Value & vv = binds.alloc(s);
```
Comment (edolstra) [inline src/libexpr/flake/flake.cc:808]:
```suggestion
    for (const auto & [key, value] : attrs) {
        auto s = state.symbols.create(key);
        auto & vv = binds.alloc(s);
```

File: src/libexpr/flake/flake.cc [hunk 7]
```diff
@@ -793,6 +793,99 @@ static RegisterPrimOp r2({
...
+        if (std::holds_alternative<std::string>(value)) {
+            vv.mkString(std::get<std::string>(value));
+        }
+        else if (std::holds_alternative<uint64_t>(value)) {
+            vv.mkInt(std::get<uint64_t>(value));
+        }
+        else { /* bool */
+            vv.mkBool(std::get<Explicit<bool>>(value).t);
+        }
+
```
Comment (edolstra) [inline src/libexpr/flake/flake.cc:818]:
```suggestion
```

File: doc/manual/src/release-notes/rl-next.md [hunk 8]
```diff
@@ -4,3 +4,7 @@
 
 - Nix now allows unprivileged/[`allowed-users`](../command-ref/conf-file.md#conf-allowed-users) to sign paths.
   Previously, only [`trusted-users`](../command-ref/conf-file.md#conf-trusted-users) users could sign paths.
+  
+- Two new builtin functions, `builtins.parseFlakeRef` and `builtins.flakeRefToString`, have been added.
+  These functions may be useful for processing `flake` `inputs`, `builtins.getFlake`, and
+  `builtins.fetchTree` style URLs/argument-sets.
```
Comment (edolstra) [inline doc/manual/src/release-notes/rl-next.md:10]:
```suggestion
  These functions are useful for converting between flake references encoded
  as attribute sets and URLs.
```

**Review by edolstra — APPROVED** [review]

File: src/libexpr/flake/flake.cc [hunk 9]
```diff
@@ -793,6 +793,94 @@ static RegisterPrimOp r2({
...
+    .name =  "__parseFlakeRef",
+    .args = {"flake-ref"},
+    .doc = R"(
+      Parse a flake reference, and return its exploded form. For example:
+
+      ```nix
+      builtins.parseFlakeRef "github:NixOS/nixpkgs/23.05?dir=lib"
+      # ==> { dir = "lib"; owner = "NixOS"; ref = "23.05"; repo = "nixpkgs";
+      #       type = "github"; }
+      ```
```
Comment (Ericson2314) [inline src/libexpr/flake/flake.cc:828]:
```suggestion
      Parse a flake reference, and return its exploded form.

      For example:
      ```nix
      builtins.parseFlakeRef "github:NixOS/nixpkgs/23.05?dir=lib"
      ```
      evaluates to:
      ```nix
      { dir = "lib"; owner = "NixOS"; ref = "23.05"; repo = "nixpkgs"; type = "github"; }
      ```
```

File: src/libexpr/flake/flake.cc [hunk 10]
```diff
@@ -793,6 +793,94 @@ static RegisterPrimOp r2({
...
+    .args = {"attrs"},
+    .doc = R"(
+      Convert a flake reference from attribute set format to URL format.
+      For example:
+
+      ```nix
+      builtins.flakeRefToString { dir = "lib"; owner = "NixOS"; ref = "23.05";
+                                  repo = "nixpkgs"; type = "github"; }
+      # ==> "github:NixOS/nixpkgs/23.05?dir=lib"
+      ```
```
Comment (Ericson2314) [inline src/libexpr/flake/flake.cc:878]:
```suggestion
      Convert a flake reference from attribute set format to URL format.

      For example:
      ```nix
      builtins.flakeRefToString {
        dir = "lib"; owner = "NixOS"; ref = "23.05"; repo = "nixpkgs"; type = "github";
      }
      ```
      evaluates to
      ```nix
       "github:NixOS/nixpkgs/23.05?dir=lib"
      ```
```

**Review by Ericson2314 — APPROVED** [review]
With those docs changes
(we don't really worry about line length in the multiline strings, but rather try to use 1 sentance per line.)
and the release notes conflicts fixed (because 2.17 was just cut) this is good!

**Comment by Ericson2314** [comment]
note the release notes still have a conflict, because 2.17 was just released.

**Comment by thufschmitt** [comment]
Coming a bit late, but from the 2023-07-24 Nix maintainers meeting:

Approved on the principle, assigned to @Ericson2314

**Pull request merged**